File: P4/ruby/BetaPowerEfficientSpaceStation.rb
# encoding: UTF-8

require_relative 'PowerEfficientSpaceStation'
require_relative 'Dice'
require_relative 'BetaPowerEfficientSpaceStationToUI'

module Deepspace
    class BetaPowerEfficientSpaceStation < PowerEfficientSpaceStation
        @@EXTRAEFFICIENCY = 1.2

        def initialize(station) # station : SpaceStation
            super
            @dice = Dice.new
        end
        
        def getUIversion
        	BetaPowerEfficientSpaceStationToUI.new(self)
        end
        
        def to_s
        	getUIversion.to_s
      	end

        def fire
            @dice.extraEfficiency ? super * @@EXTRAEFFICIENCY : super
        end
    end
end

File: P4/java/deepspace/src/deepspace/BetaPowerEfficientSpaceStation.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

/**
 *
 * @author jruib
 */
public class BetaPowerEfficientSpaceStation extends PowerEfficientSpaceStation {
    private static final float EXTRAEFFICIENCY = (float)1.2;
    
    Dice dice;
    
    BetaPowerEfficientSpaceStation(SpaceStation station) {
        super(station);
        dice = new Dice();
    }
    
    public BetaPowerEfficientSpaceStationToUI getUIversion() {
        return new BetaPowerEfficientSpaceStationToUI(this);
    }
    
    public String toString() {
        return getUIversion().toString();
    }
    
    @Override
    public float fire() {
        return dice.extraEfficiency() ? super.fire() * EXTRAEFFICIENCY: super.fire();
    }
}
